src/utils/uploader.js

```diff
@@ -0,0 +1,425 @@
+        if (status) {
+            return uploader;
+        }
+        return null;
+    }
+
+    constructor(pk, rpc, chainId, contractAddress) {
+        const provider = new ethers.JsonRpcProvider(rpc);
+        this.#wallet = new ethers.Wallet(pk, provider);
+        this.#chainId = chainId;
+        this.#contractAddress = contractAddress;
+        this.#mutex = new Mutex();
+    }
+
+    async #init(pk, rpc, contractAddress, uploadType) {
+        const status = await this.#initType(uploadType);
+        if (!status) {
+            return false;
+        }
+
+        this.#flatDirectory = await FlatDirectory.create({
+            rpc: rpc, privateKey: pk, address: contractAddress
+        });
+        return true;
+    }
+
+    async #initType(uploadType) {
+        const fileContract = new ethers.Contract(this.#contractAddress, FlatDirectoryAbi, this.#wallet);
+        let isSupportBlob;
+        try {
+            isSupportBlob = await fileContract.isSupportBlob();
+        } catch (e) {
+            console.log(`ERROR: Init upload type fail.`, e.message);
+            return false;
+        }
+
+        if (uploadType) {
+            // check upload type
+            if (!isSupportBlob && uploadType === VERSION_BLOB) {
+                console.log(`ERROR: The current network does not support this upload type, please switch to another type. Type=${uploadType}`);
+                return false;
+            }
+            this.#uploadType = uploadType;
+        } else {
+            this.#uploadType = isSupportBlob ? VERSION_BLOB : VERSION_CALL_DATA;
+        }
+        return true;
+    }
+
+    // estimate cost
+    async estimateCost(path, gasPriceIncreasePercentage) {
+        let totalFileCount = 0;
+        let totalStorageCost = 0n;
+        let totalGasCost = 0n;
+
+        const syncPoolSize = 5;
+        const gasFeeData = await this.#wallet.provider.getFeeData();
+        return new Promise((resolve, reject) => {
+            from(recursiveFiles(path, ''))
+                .pipe(mergeMap(info => this.#estimate(info, gasFeeData, gasPriceIncreasePercentage), syncPoolSize))
+                .subscribe({
+                    next: (cost) => {
+                        totalFileCount++;
+                        totalStorageCost += cost.totalStorageCost;
+                        totalGasCost += cost.totalGasCost;
+                    },
+                    error: (error) => { reject(error) },
+                    complete: () => {
+                        resolve({
+                            totalFileCount,
+                            totalStorageCost,
+                            totalGasCost,
+                        });
+                    }
+                });
+        });
+    }
+
+    async #estimate(info, gasFeeData, gasPriceIncreasePercentage) {
+        try {
+            if (this.#uploadType === VERSION_BLOB) {
+                return await this.#estimateFileByBlob(info);
```

Comment:
Why we don't consider the gasPriceIncreasePercentage when doing `estimateFileByBlob`?